internal/internal.go

```diff
@@ -612,39 +620,123 @@ func getParsedPathAndQuery(req *ext_authz.CheckRequest) ([]interface{}, map[stri
-	body := req.GetAttributes().GetRequest().GetHttp().GetBody()
+func getParsedBody(req *ext_authz.CheckRequest, parsedPath []interface{}, protoDescriptor string) (interface{}, bool, error) {
 	headers := req.GetAttributes().GetRequest().GetHttp().GetHeaders()
 
-	if body == "" {
-		return nil, false, nil
-	}
-
-	if val, ok := headers["content-length"]; ok {
-		cl, err := strconv.ParseInt(val, 10, 64)
-		if err != nil {
-			return nil, false, err
-		}
-
-		if cl != -1 && cl > int64(len(body)) {
-			return nil, true, nil
-		}
-	}
-
 	var data interface{}
 
 	if val, ok := headers["content-type"]; ok {
 		if strings.Contains(val, "application/json") {
+
+			body := req.GetAttributes().GetRequest().GetHttp().GetBody()
+
+			if body == "" {
+				return nil, false, nil
+			}
+
+			if val, ok := headers["content-length"]; ok {
+				cl, err := strconv.ParseInt(val, 10, 64)
+				if err != nil {
+					return nil, false, err
+				}
+				if cl != -1 && cl > int64(len(body)) {
+					return nil, true, nil
+				}
+			}
+
 			err := util.Unmarshal([]byte(body), &data)
 			if err != nil {
 				return nil, false, err
 			}
+		} else if strings.Contains(val, "application/grpc") {
+
+			if protoDescriptor == "" {
+				return nil, false, nil
+			}
+
+			rawbody := req.GetAttributes().GetRequest().GetHttp().GetRawBody()
+
+			if len(rawbody) <= 0 {
+				return nil, false, fmt.Errorf("invalid raw body")
+			}
+			if len(parsedPath) <= 1 {
+				return nil, false, fmt.Errorf("invalid parsed path")
+			}
+
+			err := getGRPCBody(rawbody, parsedPath, &data, protoDescriptor)
+			if err != nil {
+				return nil, false, err
+			}
 		}
 	}
 
 	return data, false, nil
 }
 
+func getGRPCBody(in []byte, parsedPath []interface{}, data interface{}, protoDescriptor string) error {
+
+	// the first 5 bytes are part of gRPC framing. We need to remove them to be able to parse
+	//https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md
+
+	in = in[5:]
+
+	bytes, err := ioutil.ReadFile(protoDescriptor) //grpcprotoset/data.pb")
+	if err != nil {
+		fmt.Println(err)
+		return err
+	}
+
+	var fileSet descriptor.FileDescriptorSet
+	if err := proto.Unmarshal(bytes, &fileSet); err != nil {
+		return err
+	}
+
+	fd, err := desc.CreateFileDescriptorFromSet(&fileSet)
+	if err != nil {
+		return err
+	}
+
+	inputType := ""
+	packageName := fd.GetPackage()
+	pathServiceRaw := parsedPath[0].(string)
+	pathService := strings.Replace(pathServiceRaw, packageName+".", "", 1)
+
+	for _, v := range fd.GetServices() {
+		if v.GetName() == pathService {
+			for _, z := range v.GetMethods() {
+				if z.GetName() == parsedPath[1] {
+					inputType = z.GetInputType().GetName()
+				}
+			}
+		}
+	}
+
+	if packageName == "" || inputType == "" {
+		return fmt.Errorf("InputType not defined")
+	}
+
+	messageName := fmt.Sprintf("%s.%s", packageName, inputType)
+
+	msgDesc := fd.FindMessage(messageName)
+
+	message := dynamic.NewMessage(msgDesc)
+
+	if err := proto.Unmarshal(in, message); err != nil {
```

Comment:
I believe instead of skipping the first few bytes ([above](https://github.com/open-policy-agent/opa-envoy-plugin/pull/202/files#diff-67213f39b84a013e7db04664cd3ad943b7a237d97b9c6e862d10806ac27def28R681)), we might be able to do something like this _here_:

```suggestion
	if _, err := pbutil.ReadDelimited(bytes.NewReader(in), message); err != nil
```

Reply:
I was wrong here, see https://github.com/open-policy-agent/opa-envoy-plugin/pull/202#discussion_r506349361